fink_filters/filter_rate_based_kn_candidates/filter.py

```diff
@@ -0,0 +1,305 @@
+# limitations under the License.
+
+from pyspark.sql.functions import pandas_udf, PandasUDFType
+from pyspark.sql.types import BooleanType
+
+import numpy as np
+import pandas as pd
+import datetime
+import requests
+import os
+import logging
+
+from astropy.coordinates import SkyCoord
+from astropy.coordinates import Angle
+from astropy import units as u
+from astropy.time import Time
+
+from fink_science.conversion import dc_mag
+
+
+@pandas_udf(BooleanType(), PandasUDFType.SCALAR)
+def rate_based_kn_candidates(
+        objectId, rfscore, snn_snia_vs_nonia, snn_sn_vs_all, drb,
+        classtar, jdstarthist, ndethist, cdsxmatch, ra, dec, ssdistnr, cjdc,
+        cfidc, cmagpsfc, csigmapsfc, cmagnrc, csigmagnrc, cmagzpscic,
+        cisdiffposc
+        ) -> pd.Series:
+    """
+    Return alerts considered as KN candidates.
+
+    If the environment variable KNWEBHOOK is defined and match a webhook url,
+    the alerts that pass the filter will be sent to the matching Slack channel.
+
+    Parameters
+    ----------
+    objectId: Spark DataFrame Column
+        Column containing the alert IDs
+    rfscore, snn_snia_vs_nonia, snn_sn_vs_all: Spark DataFrame Columns
+        Columns containing the scores for: 'Early SN Ia',
+        'Ia SN vs non-Ia SN', 'SN Ia and Core-Collapse vs non-SN events'
+    drb: Spark DataFrame Column
+        Column containing the Deep-Learning Real Bogus score
+    classtar: Spark DataFrame Column
+        Column containing the sextractor score
+    jdstarthist: Spark DataFrame Column
+        Column containing earliest Julian dates of epoch [days]
+    ndethist: Spark DataFrame Column
+        Column containing the number of prior detections (theshold of 3 sigma)
+    cdsxmatch: Spark DataFrame Column
+        Column containing the cross-match values
+    ra: Spark DataFrame Column
+        Column containing the right Ascension of candidate; J2000 [deg]
+    dec: Spark DataFrame Column
+        Column containing the declination of candidate; J2000 [deg]
+    ssdistnr: Spark DataFrame Column
+        distance to nearest known solar system object; -999.0 if none [arcsec]
+    cjdc, cfidc, cmagpsfc, csigmapsfc, cmagnrc, csigmagnrc, cmagzpscic: Spark DataFrame Columns
+        Columns containing history of fid, magpsf, sigmapsf, magnr, sigmagnr,
+        magzpsci, isdiffpos as arrays
+    Returns
+    ----------
+    out: pandas.Series of bool
+        Return a Pandas DataFrame with the appropriate flag:
+        false for bad alert, and true for good alert.
+    """
+    # Extract last (new) measurement from the concatenated column
+    jd = cjdc.apply(lambda x: x[-1])
+    fid = cfidc.apply(lambda x: x[-1])
+    isdiffpos = cisdiffposc.apply(lambda x: x[-1])
+
+    high_drb = drb.astype(float) > 0.9
+    high_classtar = classtar.astype(float) > 0.4
+    new_detection = jd.astype(float) - jdstarthist.astype(float) < 14
+    small_detection_history = ndethist.astype(float) < 20
```

Comment:
Is `small_detection_history` necessary here?

Reply:
It does remove things - on April I go from 34 to 48 candidates without it, and they have up to 70 points in their history.
They are not necessarily weird candidates, I checked a few ones and they seemed to be supernovae, but at this point it's quite late to mistake them for kilonovae.

Reply:
OK thanks! Let's keep it then